> First of all, there's a disagreement on what the variable can be:
> 
> ```
> reachability, vars:
> V1: Vehicle{type A = P} ::= Bus | Car | null// Set(Bus, Bicycle, null, Car) // = x1
> ```
> 
> The domain knows it can't be Bicycle, but the variables constants still contains the Bicycle constant.

This comes from registering the variable = constant equalities from the case defs, i.e. the actual `case Bicycle`:

```scala
      object gatherEqualities extends PropTraverser {
        override def apply(p: Prop) = p match {
          case Eq(v, NullConst) if !modelNull => vars += v // not modeling equality to null
          case Eq(v, c)                       => vars += v; v.registerEquality(c)
          case _                              => super.apply(p)
        }
      }
```

And the reason before this patch we got:

```
test/files/neg/t9657.scala:14: warning: unreachable code
    case Bus(_)  => Bus(100) // was: "unreachable code" warning 1
                       ^
```

is because there the SAT solving fails instantly:

```
reachability, vars: V1: Vehicle{type A = P} ::= null// Set(Bicycle, Car, Bus, null) // = x1
equality axioms:
  And(Set(Or(Set(V1=null#1))))
trying to reach:
  And(Set(Not(V1=null#1), V1=Bus#3))
under prefix:
  V1=null#1
  Or(Set(V1=null#1, Not(V1=Car#2)))
```

If V1 is null, can it be (non-null) Bus?  No, unreachable.

Where does `V1=null` come from?  Well it's `Or(Set(V1=null#1, Not(V1=Car#2)))` except here without the Car part, because it doesn't consider Car to be a subtype of `Vehicle { type A = P }` with the `P <: Petrol.type` skolem.

I guess at least I feel less concerned about the domain being a subset and not equal to the variable consts.  But something is going wrong as the solver is assigning V1 to a constant it can't be.

--- a/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
@@ -95,9 +95,16 @@ trait TreeAndTypeAnalysis extends Debugging {
       val parentSubtypes = tp.parents.flatMap(parent => enumerateSubtypes(parent, grouped))
       if (parentSubtypes.exists(_.nonEmpty)) {
         // If any of the parents is enumerable, then the refinement type is enumerable.
-        // We must only include subtypes of the parents that conform to `tp`.
-        // See neg/virtpatmat_exhaust_compound.scala for an example.
-        parentSubtypes.map(_.filter(_ <:< tp))
+        // We must only include subtypes of the parents that conform to `tpApprox`.
+        // See neg/virtpatmat_exhaust_compound.scala and pos/t9657.scala for examples.
+        val approximateTypeSkolemsToUpperBound = new TypeMap { // from approximateAbstracts
+          def apply(tp: Type): Type = tp.dealiasWiden match {
+            case TypeRef(_, sym, _) if sym.isTypeSkolem => tp.upperBound
+            case _                                      => mapOver(tp)
+          }
+        }
+        val tpApprox = approximateTypeSkolemsToUpperBound(tp)
+        parentSubtypes.map(_.filter(_ <:< tpApprox))
       } else Nil
     }
 

--- a/test/files/pos/t9657.scala
+++ b/test/files/pos/t9657.scala
@@ -1,0 +1,21 @@
+// scalac: -Werror
+sealed trait PowerSource
+case object Petrol extends PowerSource
+case object Pedal  extends PowerSource
+
+sealed abstract class Vehicle { type A <: PowerSource }
+case object Bicycle       extends Vehicle { type A = Pedal.type  }
+case class Bus(fuel: Int) extends Vehicle { type A = Petrol.type }
+case class Car(fuel: Int) extends Vehicle { type A = Petrol.type }
+
+object Test {
+  def refuel[P <: Petrol.type](vehicle: Vehicle { type A = P }): Vehicle = vehicle match {
+    case Car(_) => Car(100)
+    case Bus(_) => Bus(100) // was: "unreachable code" warning
+  }
+
+  def main(args: Array[String]): Unit = {
+    println(refuel(Car(100)))
+    println(refuel(Bus(5)))
+  }
+}

